### ISSUE TITLE
Too slow on GPU

### ISSUE THREAD

**Issue by umgupta** [deepmind/optax#319]
I ran the [flax example](https://github.com/deepmind/optax/blob/master/examples/flax_example.py) from the examples folder on GPU and CPU. 

It runs lightning fast on the CPU taking less than 2s, whereas on GPU it took more than 28s. I observed the same problem in another training code. Infact, that is where I first figured that the optax specific part is slow/bottleneck. 

Is there any reason why this is happening and how to fix it?

If it helps to know, I am using `jax==0.3.4` installed via `pip install jax[cuda]` and `optax==0.1.0`, and `cudatoolkit==11.3.1`

Thanks,

**Comment by mkunesch** [comment]
Hi! Oh, interesting - thanks for raising this! I took a look at the examples and the behaviour you describe could be due to the placement of `jax.jit`.

The `jit` is only placed around the loss function and not the optax calls. This means that the optax code will run slowly and not on device. For an example on where to place the jit to get the best performance, see the lookahead example. We should definitely change the flax example to include the optax update and apply in the `jit` too.

In the haiku example and the quick start there is no `jax.jit` at all - presumably on purpose, but I think it might be good to add it to demonstrate the best practice of placing the jit around the optax calls too.

Would this explain the performance issue in the other training code you mentioned too?

**Comment by umgupta** [comment]
> Would this explain the performance issue in the other training code you mentioned too?

Yes, this explains the issue. Sorry for raising a bug without trying this --- I feel stupid.

> Would you like to file the PR for making the changes to where `jit` is used? No worries if not, someone else can jump in too or I'm also happy to make the change.



> Would this explain the performance issue in the other training code you mentioned too?

Yes, this worked. I am sorry for raising an issue without trying jitting --- I thought optax would do it already for me.  

> Would you like to file the PR for making the changes to where `jit` is used? No worries if not, someone else can jump in too or I'm also happy to make the change.

I would be happy to fix this. Did you mean for me to add `jit` in the examples or all the optimizers?

**Comment by mkunesch** [comment]
Hi! Not at all - thanks for raising this! I think the placement of the jit in the flax example isn't the best so I think we should change this.

Regarding optax including the jit around the optimizers: we don't do this for several reasons (it would almost always be the wrong place for the jit, user might want different transformations or different arguments for the jit, and in some cases users might not want the jit at all). In most cases, it is best to place the jit around the outermost jittable function. In the case of neural network training this is usually the function that takes one training step.

I think we should change the flax example to have the jit around an entire training step and not just the loss (similar to how the lookahead example does it).

Thanks a lot!
